File: app/src/main/java/com/zibete/proyecto1/Adapters/FavoritesDiffCallback.java
package com.zibete.proyecto1.Adapters;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import com.zibete.proyecto1.MainActivity;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class FavoritesDiffCallback extends DiffUtil.Callback {

    private ArrayList<String> mNewList;
    private ArrayList<String> mOldList;


    public FavoritesDiffCallback(List<String> newlist, List<String> oldlist) {
        this.mNewList = (ArrayList<String>) newlist;
        this.mOldList = (ArrayList<String>) oldlist;
    }

    @Override
    public int getOldListSize() {
        return mOldList.size();
    }

    @Override
    public int getNewListSize() {
        return mNewList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {




        return mOldList.get(oldItemPosition).getBytes()==mNewList.get(newItemPosition).getBytes();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition).equals(mNewList.get(newItemPosition));
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        String newItem = mNewList.get(newItemPosition);
        String oldItem = mOldList.get(oldItemPosition);

        Bundle diff = new Bundle();


        if(!Objects.equals(newItem.getBytes(), oldItem.getBytes())){
            diff.putString("id", Arrays.toString(newItem.getBytes()));
        }


        if (diff.size()==0){
            return null;
        }
        return diff;
    }
}

File: app/src/main/java/com/zibete/proyecto1/adapters/FavoritesDiffCallback.kt
package com.zibete.proyecto1.adapters

import androidx.recyclerview.widget.DiffUtil
import com.zibete.proyecto1.ui.favorites.FavoriteUserUi

object FavoritesDiffCallback : DiffUtil.ItemCallback<FavoriteUserUi>() {

    override fun areItemsTheSame(oldItem: FavoriteUserUi, newItem: FavoriteUserUi): Boolean =
        oldItem.id == newItem.id

    override fun areContentsTheSame(oldItem: FavoriteUserUi, newItem: FavoriteUserUi): Boolean =
        oldItem == newItem

    override fun getChangePayload(oldItem: FavoriteUserUi, newItem: FavoriteUserUi): Any? {
        val changed = mutableSetOf<String>()

        if (oldItem.isOnline != newItem.isOnline) changed += PayloadFavoriteUser.IS_ONLINE
        if (oldItem.profilePhoto != newItem.profilePhoto) changed += PayloadFavoriteUser.PHOTO_URL
        if (oldItem.name != newItem.name) changed += PayloadFavoriteUser.NAME
        if (oldItem.age != newItem.age) changed += PayloadFavoriteUser.AGE

        return changed.takeIf { it.isNotEmpty() }
    }

    object PayloadFavoriteUser {
        const val IS_ONLINE = "payload_favorite_user_is_online"
        const val PHOTO_URL = "payload_favorite_user_photo_url"
        const val NAME = "payload_favorite_user_name"
        const val AGE = "payload_favorite_user_age"
    }
}



